Direct planner

YLGH opened the pull request:
Summary:
See  https://docs.google.com/document/d/1ASiz0Fu3c8cCHmsbseUdulU4jfTgdwZAxvlywgXiBAU/edit for more context

tl;dr:

Currently the only “feasible” way to specify a sharding plan is by using the OSS planner, and using constraints. I say “feasible” because although we can manually construct a ShardingPlan/ParameterSharding, it’s quite cumbersome to do so. Additionally, users do not have the expressibility to say “I want to place this parameter on rank 0” when using planner.

With this new API we can construct sharding plans in a direct way:

```
ebc = EmbeddingBagCollection(...)
plan = construct_parameter_sharding_plans(
            ebc,
            {
                "table_0": data_parallel_sharding(),
                "table_1": row_wise_sharding(),
                "table_2": column_wise_sharding(),
                "table_3": column_wise_sharding(ranks=[0,1,2]),
                "table_4": table_row_wise_sharding(host_index=2),
            },
            EmbeddingBagCollectionSharder()
        )
ebc = sharder.shard(
   ebc,
   plan
)
```

Note, that this is not intended to replace the greedy planner. Instead it allows users to shard modules without being FORCED to use planner. Users will be able to user planner in the exact same way.

Having alternatives instead of planner has a couple of benefits

1. There are bugs in the planner (we should be able to shard something, but planner may have a false negative rejection)
2. Planner doesn't correctly account for new architectures/models /optimizers
3. Easier to directly use torchrec (new users don't need to understand how to use planner).

-----

So I decided to use a generator function because alternative ways would require creating a new dataclass called like TableWiseConfig(row=...), and I didn't like the look of it.

Having just the atomic generator functions is also nice since it doesn't couple these configs with the construct_parameter_sharding_plans

Differential Revision: D39803183

divchenko commented on torchrec/distributed/planner/tests/test_simple_planner.py:
```diff
@@ -0,0 +1,232 @@
+    world_size: int,
+    kjt_input_per_rank: List[KeyedJaggedTensor],
+    sharder: ModuleSharder[nn.Module],
+    backend: str,
```
nit: group rank, world_size, local size, backend together . Maybe introduce a dataclass so it's easier to pass all the way to MultiProcessContext ?

divchenko commented on torchrec/distributed/planner/tests/test_simple_planner.py:
```diff
@@ -0,0 +1,232 @@
+        sharded_model_pred_kt = sharded_model(kjt_input_per_rank[ctx.rank]).to_dict()
+        _sharded_model_pred = torch.stack(  # noqa
+            [sharded_model_pred_kt[feature] for feature in feature_keys]
+        )
```
Do you intend to check something here? e.g. sharded vs unsharded results?

YLGH replied:
here I just wanted to check that things run w/ the constructed plans,

there will be another unit test that does correctness validation over composable ebc with inputs from this

divchenko commented on torchrec/distributed/planner/planners.py:
```diff
@@ -302,3 +300,267 @@ def plan(
+            if param_name in name:
+                param = _param
+                break
+        assert param is not None
```
better error message

YLGH replied:
the param_name in name is also a bit annoying because we strip off ".weight", I guess I can get substring bugs

divchenko commented on torchrec/distributed/planner/planners.py:
```diff
@@ -302,3 +300,267 @@ def plan(
+    module: nn.Module,
+    per_param_sharding: Dict[
+        str, Callable[[nn.Parameter, int, int, str], ParameterSharding]
+    ],
```
This Callable is complex and repeated in many places. Define a new type.
ParameterShardingGenerator = Callable[[nn.Parameter, int, int, str], ParameterSharding]

divchenko commented on torchrec/distributed/planner/planners.py:
```diff
@@ -302,3 +300,267 @@ def plan(
+    return per_parameter_sharding
+
+
+def uniform_sharding(
```
What is the final end-to-end code? We need to list an example somewhere.
```
model = MyModel()
ebc_plan = construct_parameter_sharding_plans(model.ebc, uniform_sharding(model.ebc, row_wise_sharding())
eb_plan = construct_parameter_sharding_plans(model.eb, uniform_sharding(model.eb, row_wise_sharding())
plan = {**eb_plan, **eb_plan}
```

YLGH replied:
yeah 
I think there needs to be something that can add FQN path prefix to output of construct_parameter_sharding_plans: Dict[str, ParameterSharding], 

this operates on the Sharder.shard level

YLGH replied:
```
Overall plan would look like
{
   "model.ebc": ebc_plan,
   "model.ec" : ec_plan
}
``` i suppose

divchenko commented on torchrec/distributed/planner/planners.py:
```diff
@@ -302,3 +300,267 @@ def plan(
+    per_param_sharding: Dict[
+        str, Callable[[nn.Parameter, int, int, str], ParameterSharding]
+    ],
+    sharder: ModuleSharder[nn.Module],
```
We need to add a higher-level API, where we don't pass sharder. There is a module -> sharder dict. It should operate on top of it.

YLGH commented on torchrec/distributed/planner/planners.py:
```diff
@@ -302,3 +300,267 @@ def plan(
+    parameter_sharding_generator: Callable[
+        [nn.Parameter, int, int, str], ParameterSharding
+    ],
+    sharder: ModuleSharder[nn.Module],
```
default to None and try to find the correct one from dict

colin2328 commented on torchrec/distributed/planner/planners.py:
```diff
@@ -302,3 +300,267 @@ def plan(
+        ),
+        sharding_type=sharding_type,
+        # pyre-ignore
+        compute_kernel=None,
```
setting compute_kernel to None will mean that we cannot use w/ DMP (unless manually changed), IMO add comment here or in user facing API

colin2328 commented on torchrec/distributed/planner/planners.py:
```diff
@@ -302,3 +300,267 @@ def plan(
+
+
+def column_wise_sharding(
+    shard_dim: int = 128,
```
add docstrings, including what shard_dim represents

colin2328 commented on torchrec/distributed/planner/planners.py:
```diff
@@ -302,3 +300,267 @@ def plan(
+        that takes in [parameter, local_size, world_size, device_type] and returns a ParameterSharding. We recommend using one of the predefined generator functions
+        e.g. table_wise_sharding, row_wise_sharding, etc,
+        sharder: ModuleSharder[nn.Module]: Sharder that we are creating a plan for,
+        local_size: Optional[int] = None: process group local size
```
add generally user should not pass these in

colin2328 commented on torchrec/distributed/planner/planners.py:
```diff
@@ -302,3 +300,267 @@ def plan(
+        sharder: ModuleSharder[nn.Module]: Sharder that we are creating a plan for,
+        local_size: Optional[int] = None: process group local size
+        world_size: Optional[int] = None: process_group world_size
+        device_type: str : Torch device type,
```
IMO we should use device, or at least torch.device.type rather than str

colin2328 commented on torchrec/distributed/planner/planners.py:
```diff
@@ -302,3 +300,267 @@ def plan(
+                "table_0": data_parallel_sharding(),
+                "table_1": row_wise_sharding(),
+                "table_2": column_wise_sharding(),
+                "table_3": column_wise_sharding(ranks=[0,1,2]),
```
consider allowing this, or if not at least update to reflect (need to specify shard_dim)

colin2328 commented on torchrec/distributed/planner/planners.py:
```diff
@@ -302,3 +300,267 @@ def plan(
+                "table_1": row_wise_sharding(),
+                "table_2": column_wise_sharding(),
+                "table_3": column_wise_sharding(ranks=[0,1,2]),
+                "table_4": table_row_wise_sharding(host_index=2),
```
we need to write it or comment that it's not supported

colin2328 commented on torchrec/distributed/planner/planners.py:
```diff
@@ -302,3 +300,267 @@ def plan(
+    for param_name, sharding_plan_generator in per_param_sharding.items():
+        param = None
+        for name, _param in module.named_parameters():
+            if param_name in name:
```
we need to strip weight and actually match rather than relying on substring match

colin2328 replied:
since you already do assertion above just iterate through and do .get()

colin2328 commented on torchrec/distributed/planner/planners.py:
```diff
@@ -302,3 +300,267 @@ def plan(
+    Example::
+    ebc = EmbeddingBagCollection(...)
+    sharder = EmbeddingBagCollectionSharder()
+    plan = construct_parameter_sharding_plan(
```
IMO we should avoid uniform and instead call it something that indicates were sharding all tables like this "shard_all_tables"

Commit:
Direct planner (#662)

Summary:
Pull Request resolved: https://github.com/pytorch/torchrec/pull/662

See  https://docs.google.com/document/d/1ASiz0Fu3c8cCHmsbseUdulU4jfTgdwZAxvlywgXiBAU/edit for more context

tl;dr:

Currently the only “feasible” way to specify a sharding plan is by using the OSS planner, and using constraints. I say “feasible” because although we can manually construct a ShardingPlan/ParameterSharding, it’s quite cumbersome to do so. Additionally, users do not have the expressibility to say “I want to place this parameter on rank 0” when using planner.

With this new API we can construct sharding plans in a direct way:

```
ebc = EmbeddingBagCollection(...)
plan = construct_parameter_sharding_plans(
            ebc,
            {
                "table_0": data_parallel_sharding(),
                "table_1": row_wise_sharding(),
                "table_2": column_wise_sharding(),
                "table_3": column_wise_sharding(ranks=[0,1,2]),
                "table_4": table_row_wise_sharding(host_index=2),
            },
            EmbeddingBagCollectionSharder()
        )
ebc = sharder.shard(
   ebc,
   plan
)
```

Note, that this is not intended to replace the greedy planner. Instead it allows users to shard modules without being FORCED to use planner. Users will be able to user planner in the exact same way.

Having alternatives instead of planner has a couple of benefits

1. There are bugs in the planner (we should be able to shard something, but planner may have a false negative rejection)
2. Planner doesn't correctly account for new architectures/models /optimizers
3. Easier to directly use torchrec (new users don't need to understand how to use planner).

-----

So I decided to use a generator function because alternative ways would require creating a new dataclass called like TableWiseConfig(row=...), and I didn't like the look of it.

Having just the atomic generator functions is also nice since it doesn't couple these configs with the construct_parameter_sharding_plans

Differential Revision: D39803183

fbshipit-source-id: ef4b45cd6118a3ccab786f128050f45b50d1bfe4
